python/tvm/auto_scheduler/cost_model/xgb_model.py

```diff
@@ -0,0 +1,590 @@
+        self.context_dict = defaultdict(dict)
+
+    def get(self, key, matrix, default=None):
+        """
+        Get an attribute of a xgb.DMatrix
+
+        Parameters
+        ----------
+        key: str
+            The name of the attribute
+        matrix: xgb.DMatrix
+            The matrix
+        default: Optional
+            The default value if the item does not exist
+        """
+        return self.context_dict[key].get(matrix.handle.value, default)
+
+    def set(self, key, matrix, value):
+        """
+        Set an attribute for a xgb.DMatrix
+
+        Parameters
+        ----------
+        key: str
+            The name of the attribute
+        matrix: xgb.DMatrix
+            The matrix
+        value: Optional
+            The new value
+        """
+        self.context_dict[key][matrix.handle.value] = value
+
+dmatrix_context = XGBDMatrixContext()
+
+
+class XGBModel(PythonBasedModel):
+    """Train a XGBoost model to predict the normalized throughputs of programs.
+
+    Let the normalized throughput be the score of a program (higher is better). We predict
+    (approximiate) the score of a program = the sum of the scores of all stages in this program.
+    i.e. score(P) = score_s0 + score_s1 + ... + score_sn,
+    where score_si is the score of Stage i in Program P.
+
+    We extract feature for each stage and let the xgboost predict the score for each stage.
+    We then sum up the predictions as the score of the whole program.
+
+    We use RMSE as the loss function.  i.e. loss(P, y) = 1/2 * (score(P) - y)^2,
+    where P is the program and y is the normalized throughput according to
+    the ground truth (measurement).
+    XGBoost does not support this loss function because `score(P)` is a sum of the prediction
+    of several samples, so we implemented a custom loss function and call it pack-sum-rmse.
+    It is called "pack-sum" because we combine several samples into a "pack" and sum up
+    their predictions.
+    """
+    def __init__(self, verbose_eval=25, num_warmup_sample=100, seed=None):
+        self.xgb_params = {
+            'max_depth': 10,
+            'gamma': 0.001,
+            'min_child_weight': 0,
+            'eta': 0.2,
+            # todo(merrymercy): automatically decrease learning rate when the loss is too large
+
+            'n_gpus': 0,
+            'nthread': multiprocessing.cpu_count() // 2,
+            'verbosity': 0,
+            'seed': seed or 43,
+            'disable_default_eval_metric': 1
+        }
+        self.bst = None
+        self.plan_size = 32
+        self.num_warmup_sample = num_warmup_sample
+        self.verbose_eval = verbose_eval
+
+        super().__init__()
+
+        # cache measurement input/result pairs and extracted features
+        self.inputs = []
+        self.results = []
+        self.inputs_feature_cache = []
+
+    def update(self, inputs, results):
+        """Update the cost model according to new measurement results (training data).
```

Comment:
1. It would be better to mention we will train a new model from scratch every time `update` is invoked.
2. This function assumes that there won't be duplicated inputs because we guarantee the same schedule will be measured only once. It would be better to mention it as well.